### DISCUSSION TITLE
SAR aircraft into BaseStation format

### DISCUSSION THREAD

**Discussion by flygfantast** [jvde-github/AIS-catcher#73] [Ideas]
I'm using Virtual Radar Server to display my decoded ADS-B aircrafts and a nice feature would be to see SAR AIS-equipped aircraft there.

Would it be possible to add a network output in the BaseStation format for just the AIS messages regarding SAR aircraft?

http://woodair.net/sbs/article/barebones42_socket_data.htm
https://forum.virtualradarserver.co.uk/viewtopic.php?t=360

**Comment by jvde-github** [comment]
Hello, I don't have any of these tools installed myself so not very familiar but I have seen this request more often. It is technically possible but as the forum discussion highlights not all information in a BaseStation message  is available in Message 9 (i.e. planes are identified by MMSI). So, this requires some experimentation on how to map and what works for the receiving software. 
In principle the correct route would be to implement AIS MSG 9 functionality in the ADSB plotting software. 

Nevertheless, I am more than happy to include this but I need to have some sort of mapping, example or reference implementation as a specification.

**Reply by flygfantast** [reply]
Unfortunally my knowledge is very slim. I'll try to find some more info.

**Comment by flygfantast** [comment]
I don't know if it's to any help, but the source code to VRS are on Github: https://github.com/vradarserver/vrs

**Comment by jvde-github** [comment]
I will add it to the list of things to look at. Thanks!

**Comment by flygfantast** [comment]
MMSI -> ICAO would easiest be mapped as the last 6 digits from the right. For example MMSI 111265591 -> 265591 (ICAO)

**Comment by jvde-github** [comment]
Will have a look today. Question, are you running AIS-catcher on Linux?  

Do you have a few NMEA AIS lines for a SAR objects that you can share, makes it easier with testing otherwise have to wait until a helicopter is out. 

First time I see VRS and it is quite nice!

Some experimentation with Virtual Radar Server shows that this simple format is picked up and shown on the map without confusion with existing planes:

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/52420030/219864627-c9aa404b-2308-4a2c-8f11-c377781bb1dd.png">

**Reply by flygfantast** [reply]
I'm running Windows unfortunally.

I have no saved NMEA lines, I will try to get some on monday since the military helicopters I see isn't flying on weekends. 

Yeah, VRS is a nice bit of software. :)

Looks nice :)

**Reply by jvde-github** [reply]
Ok, I made a PoC which seems to be working in VRS: 

https://github.com/jvde-github/ais2adsb

The program listens to NMEA lines over UDP and sends BaseStation messages to a TCP server like VRS. 
If this works well, and has converged to something useful, I can also integrate in AIS-catcher.

This is an example where I also include ships in the BaseStation output (as I was waiting for SAR helicopters):

![image](https://user-images.githubusercontent.com/52420030/219868547-0071c93d-475e-4295-89de-ee8e3d01e321.png)

In principle this should run on both Windows as Linux but requires Python and pyais to be installed.

**Reply by flygfantast** [reply]
Nice, I will try to get it running on my Windows machine.

**Reply by flygfantast** [reply]
Got it running without any problems. Unfortunally the traffic is VERY low in the winter at my place, so only stationary sailingships and a basestation in my area. Guess it needs to be moving for the message to be passed on to VRS. Hopefully the military helicopters with AIS transponders will be out tomorrow.

**Reply by jvde-github** [reply]
Great! I think to make it useful we need to have a good mapping of MMSI codes to ICAO mode S codes. The program already includes a table that was provided to me by another user (who was mentioning good results in VRS for SAR via AIS vs ADS-B so that is encouraging). 
If you have any MMSIs for SAR aircraft in your region with corresponding ICAO codes, let me know and I can add.

I also added a standalone package for windows that should be easier to install and run although still requires the command line: https://github.com/jvde-github/ais2adsb/releases/tag/v0.1

ps.  did you add 5 parameters on the command line with a 1 as last? This will include AIS msg types 1,2 and 3 in the output. So higher chance of seeing something coming through which might be useful for testing the connection. Class B ship positions are not included in the output.

**Reply by flygfantast** [reply]
Yes, a mapping feature is great. Unfortunally the ones I have are assigned to military helicopters so no known ICAO is available for them as they never have their Mode-S transponder on.

Yes, I added the last parameter, although I got no positions right now since alt the boats are anchored.

I will give you some NMEA lines later today if the helicopters are out (but it's bad weather today, heavy snow and windy).

**Comment by flygfantast** [comment]
[111265534.txt](https://github.com/jvde-github/AIS-catcher/files/10785610/111265534.txt)

Attached is todays traffic from SAR helicopter with MMSI 111265534.

Examples:
```
!AIVDM,1,1,,A,91b76wP00K15iKPQRrF:Ld0000S:,0*1B
!AIVDM,1,1,,A,91b76wP00M15h0pQRtC:8E000@=p,0*45
!AIVDM,1,1,,A,91b76wP00S175kPQS64=1pP000SC,0*3D
!AIVDM,1,1,,A,91b76wP00c15goPQRv@:opP00@F8,0*2B
!AIVDM,1,1,,A,91b76wP00h15kB@QS3E8wV@000S:,0*1B
!AIVDM,1,1,,A,91b76wP00i177;0QS=6;jsP005bh,0*7D
!AIVDM,1,1,,A,91b76wP00j177J`QS=6;os0008KA,0*37
!AIVDM,1,1,,A,91b76wP00k15kK`QS4791mP008?1,0*47
!AIVDM,1,1,,A,91b76wP00q1786HQRih=hv0005bd,0*79
!AIVDM,1,1,,A,91b76wP00r15fjhQRnL79qP00D01,0*33
!AIVDM,1,1,,A,91b76wP00v16w7HQJh5:gm0000SA,0*77
!AIVDM,1,1,,A,91b76wP012178`hQRT5=Vph005bd,0*4D
!AIVDM,1,1,,A,91b76wP01416o:0QSb0;0h000<02,0*00
!AIVDM,1,1,,A,91b76wP01516dD@QVi>9GM@000S:,0*43
!AIVDM,1,1,,A,91b76wP01716sM0QJnf;TJh000S:,0*52
!AIVDM,1,1,,A,91b76wP01716sq8QJmS;OJ@00HIc,0*58
!AIVDM,1,1,,A,91b76wP017179khQRWV=JVP008AI,0*23
```

**Comment by jvde-github** [comment]
Thanks, tested and it worked well. I read in your file and send it to ais2adsb using the following command:
```
AIS-catcher -r txt 111265534.txt -u 192.168.1.235 4005
```
And send it as ADSB msg to VRS as follows:
```
ais2adsb 192.168.1.235 4005 192.168.1.239 30003
```
and it shows up in VRS (!!!):
<img width="197" alt="image" src="https://user-images.githubusercontent.com/52420030/220182695-00e7a845-e6f9-411c-86a9-9184ed33cc18.png">

Test passed. Many thanks for sharing.

**Reply by flygfantast** [reply]
Gr8 ;)

**Comment by flygfantast** [comment]
111265103 = 4AB423 (SE-MAC, Swedish Coast Guard KBV503)

**Reply by jvde-github** [reply]
Thanks I have added it to the ais2adsb default list.

**Comment by Mac365** [comment]
Hi Jasper, hoping all is well with you.
I have AIS-Catcher running on a raspberry pi5 & Virtual Radar Server running on a Windows10 PC both are without issues.
I'm based in London Uk around 2.5 miles from the River Thames & around 20 miles from Heathrow airport directly under the main flight path so have live ais & adsb data decoding 24/7's.
I'm using ais2adsb to view the marine traffic along with the adsb traffic using vrs. I run ais2adsb on the pi5 where ais-catcher is also running & it works well, however I'd like to have boat/ship markers for the ais traffic as it is difficult distinguishing the different traffic when its busy/ during the day.
![Vrs adsb   ais traffic](https://github.com/user-attachments/assets/1b711c31-304f-4977-93d0-1664bdcdbfa8)
Jasper are there any plans to add boat/ship markers? Maybe that has to be done on the vrs side, vrs does have a custom content plugin where code can be injected into the website, It should be possible using custom content.
I'm not too knowledgeable with code so hope  you  can  put together something that can  add boat/ship markers.
Great piece of software! Thanks it's much appreciated.

**Reply by jvde-github** [reply]
Hello, what you can do is run AIS-catcher on the Windows machine and then connect it both to the ADSB as AIS feed on the Raspberry Pi and view both on one map. Or if you run the AIS-catcher web viewer on the RPi you can include the ADSB there.
Let me know if that is of use.

I don't have any experience with running VRS and no plans to start developing for that platform. But I have seen some discussions about boat markers. Not sure if that went anywhere. Let me check with some folks.

Cheers, Jasper

**Reply by jvde-github** [reply]
<img width="1190" alt="image" src="https://github.com/user-attachments/assets/b58bce8c-afd0-4bdf-8be3-0a8e8468288d" />
In case you have not seen. AIS-catcher with ADSB feed included.

**Reply by Jonboy1081** [reply]
Hi, you can use the custom markers script with VRS, this will then give you ship markers as per the below. If you want to take things further you can map the MMSI to a hex code within AIS2ADSB and then add that hex code to your VRS basestation file, the custom markers will also show different colours for police boats, RNLI, etc if you have these mapped and in the SQB.

https://github.com/rikgale/VRSCustomMarkers

![Image](https://github.com/user-attachments/assets/3b1f695c-ae2d-4a4f-a6df-a70fa2da1a2e)

**Reply by Gazman429** [reply]
I run ais2adsb to get AIS SAR aircraft that ADS-B has either lost contact or not in range to display on VRS, works for me, minor code addition to display vessel name and SAR aircraft with known fixed ADS-B call signs.
Down side for me is where I am I don't get many SAR aircraft but what I have gotten I noticed an altitude difference between AIS and ADS-B as here AIS SAR aircraft transmit altitude in meters not feet so I see jumps in altitude as VRS switch's between last receiver hoping to fine tune it but it's a waiting game for me and usually I'm at work when one shows up.
Also been playing with AIS Catcher with ADS-B side of things, playing with a plugin for aircraft side of things to see if I can add different aircraft icons for the different aircraft.

![ais2adsb 5](https://github.com/user-attachments/assets/db8c0435-c74d-48a8-8bef-17151732ef80)
![ais2adsb 6](https://github.com/user-attachments/assets/f2211942-8b37-477f-93cc-c7ab02284624)
![ais2adsb 7](https://github.com/user-attachments/assets/3380bcb2-657f-40fc-9d1e-078d1418a8b7)

**Reply by jvde-github** [reply]
Thanks both. @Gazman429, is this something we need to fix in ais2adsb, i.e. add a conversion from meter to feet when generating the basestation output?

**Reply by Gazman429** [reply]
I've written a conversion code line in to see if it works, but for me it's a waiting game for SAR aircraft.
If there are any following this from Sydney or Melbourne would be kind enough to share AIS data with me where I know there are plenty of AIS equipped aircraft to help me to see it my code works would be great.

**Reply by Gazman429** [reply]
I've added my variation of the original code if anyone want's to use it.

https://github.com/jvde-github/ais2adsb/issues/19
